### PR TITLE
[WIP] Fix scheduler behavior for top-level lists

### LINKED ISSUE #1192: Scheduler doesn't compute lists of tasks
Lists of tasks with no external dependencies aren't computed:

```python
In [1]: from dask import get

In [2]: def inc(x):
   ...:     return x + 1
   ...:

# This doesn't work
In [3]: dsk = {'a': [1, (inc, 1)]}

In [4]: get(dsk, 'a')
Out[4]: [1, (<function __main__.inc>, 1)]

# But this does
In [5]: dsk = {'a': ['b', (inc, 1)], 'b': 1}

In [6]: get(dsk, 'a')
Out[6]: [1, 2]

# And so does this
In [7]: dsk = {'a': [1, (inc, 'b')], 'b': 1}

In [8]: get(dsk, 'a')
Out[8]: [1, 2]
```

Discovered while investigating http://stackoverflow.com/questions/37535160/using-dask-multi-threaded-module.

### PR DESCRIPTION
**Pull request by jcrist**
Previously lists were treated differently depending on if they were a toplevel key in the graph or not. The new behavior treats them the same as long as they're not nested lists. This is best illustrated with an
example:

```python
dsk = {'a': [1, 2, 3],
       'b': 'a',
       'c': [(inc, 1), 2],
       'd': [[1, 2, (inc, 1)], 4],
       'e': [(list, [1, 2, (inc, 1)]), 4],
       'f': ['a', 'b']}
```

In this example:
- 'a' is data, as none of the elements are keys or tasks
- 'b' is not data, it's just an alias to 'a' (a trivial task)
- 'c' is not data, as it contains a task
- 'd' is data, as the list contains no tasks.
- 'e' is not data, as it contains tasks. Note that the result of this is
  ``[[1, 2, 3], 4]``, which is probably the desired result for `d`.
- 'f' is not data as it contains keys in the graph

To summarize:

A value in the graph is computed if:
- It's a task
- It's a key in the graph
- It's a list that contains either tasks or keys

Note that this definition is not fully recursive, as lists of lists aren't traversed. It could be, but I didn't go that far yet.

Fixes #1192.

Todo:
- [x] decide on what the spec is
- [x] fix dask.core.get to match accordingly
- [x] figure out why the `GetFunctionTestCase` runs 3 times for dask.core.get, and stop that
- [x] document spec more thoroughly

### CODE REVIEW

**Comment by mrocklin** [comment]
I'm tempted by the option to go fully recursive.  It's simple and fully correct, even if sometimes slow.  It would be interesting to see how slow it is on large lists of small data elements and how often these come up (`db.from_sequence`?)

**Comment by jcrist** [comment]
The behavior implemented here is consistent with the existing scheduler (which traverses lists inside of tasks), except for the case of top-level lists (which it missed). Adding in recursive traversals of lists could be done, and I don't think it would slow anything down. `db.from_sequence` results in tuples of data, so that doesn't come in to play here at all.

**Comment by mrocklin** [comment]
Are there reasons to avoid fully recursive checks?

**Comment by jcrist** [comment]
Not really, it's just more different than the small change made here. Can fix. Also, is there a good reason to keep around `dask.core.get`? I'd have to update that scheduler as well.

**Comment by mrocklin** [comment]
It's nice to keep around as a reference but serves no practical value.  I would update it if it will take less than an hour.  Alternatively you could ask @eriknw .

**Comment by eriknw** [comment]
Can you fill me in on some context?  What problem is this meant to solve?

Why not just make tasks explicit as follows, which doesn't require a change to the protocol:
```python
>>> from dask.core import get
>>> inc = lambda x: x+1
>>> dsk = {'a': [1, 2, 3],
...        'b': (list, 'a'),
...        'c': (list, [(inc, 1), 2]),
...        'd': (list, [[1, 2, (inc, 1)], 4]),
...        'e': (list, [(list, [1, 2, (inc, 1)]), 4]),
...        'f': (list, ['a', 'b'])}
>>> 
>>> get(dsk, 'a')
[1, 2, 3]
>>> get(dsk, 'b')
[1, 2, 3]
>>> get(dsk, 'c')
[2, 2]
>>> get(dsk, 'd')
[<tupleiterator object at 0x101700350>, 4]
>>> list(get(dsk, 'd')[0])
[1, 2, 2]
>>> get(dsk, 'e')
[[1, 2, 2], 4]
>>> get(dsk, 'f')
[[1, 2, 3], [1, 2, 3]]
```

**Comment by jcrist** [comment]
The issue isn't that things can't be worked around using the current spec, it's that there's inconsistent behavior between top-level lists, and lists as arguments. Also the bug from the referenced issue (https://github.com/dask/dask/issues/1192#issue-157705647) is definitely a bug, even without changes in the spec. I'm unsure if the spec should be changed, but it's cheap-ish to do so, and shouldn't degrade performance significantly.

**Comment by eriknw** [comment]
I see, so that bug is because `dask.async.get_sync` doesn't follow the current spec (i.e., resolve an item iff it's a task, otherwise treat it as a literal).  It sometimes resolves top-level lists.

Instead of making tasks explicit as I did above, one could make literals explicit:
```python
>>> from dask.core import get
>>> from toolz import first
>>> 
>>> def literal(val):
...     return (first, (val,))
... 
>>> def double(x):
...     return x + x
... 
>>> dsk = {
...     'a': 1,
...     'b': (double, 'a'),
...     'c': (double, literal('a')),
...     'd': literal('c'),
... }
>>> 
>>> get(dsk, 'a')
1
>>> get(dsk, 'b')
2
>>> get(dsk, 'c')
'aa'
>>> get(dsk, 'd')
'c'
```
Call me old fashioned, but resolving top-level lists and aliases still feels kind of dirty to me IMHO.  If y'all want to make a change, though, I won't fight tooth-and-nail.  The principle of least surprise is great to follow, but different folk are surprised by different things.  Personally, I find the current spec simple, not surprising, and not scary.

**Comment by eriknw** [comment]
A compelling reason to change the spec would be if doing so makes the distributed scheduler and its futures API simpler or more efficient.

**Comment by eriknw** [comment]
Jim pointed out to me that optimizations such as fusing can cause problems when top-level lists are treated differently than lists in tasks.  For example,
```python
>>> import dask
>>> 
>>> dsk = {'a': 1, 'b': 2, 'c': ['a', 'b'], 'd': (list, 'c')}
>>> dask.get(dsk, 'd')
[1, 2]
>>> dask.core.get(dsk, 'd')
['a', 'b']
>>> 
>>> dsk_fused, _ = dask.optimize.fuse(dsk)
>>> dask.get(dsk_fused, 'd')
[1, 2]
>>> dask.core.get(dsk_fused, 'd')
[1, 2]
```
This is probably sufficient reason to update the protocol.

I can update `dask.core.get` if doing so is worthwhile.  I can also change its laziness so it returns a list:
```python
>>> import dask
>>> from toolz import identity
>>> dsk = {'a': 1, 'b': 2, 'c': (identity, ['a', 'b'])}
>>> 
>>> dask.get(dsk, 'c')
[1, 2]
>>> dask.core.get(dsk, 'c')  # change this to return a list
<tupleiterator object at 0x101700350>
>>> list(dask.core.get(dsk, 'c'))
[1, 2]
```

**Comment by jcrist** [comment]
I just pushed the fix for recursive definition. I can fix the `dask.core.get` function as well, unless you want to take it @eriknw.

**Comment by eriknw** [comment]
oh, by all means, you can tackle `dask.core.get` if you want.

**Comment by jcrist** [comment]
This took longer than it should have, but it's fixed. I had to change the interface, but since it wasn't used anywhere I don't think that's a problem. Still has both recursive and non-recursive versions. If people are ok with this, all that's left is to update the spec docs.

**Comment by jcrist** [comment]
Ping @eriknw, @mrocklin for quick review. If this spec definition/fix is good, I just need to add docs on the new behavior.

File: dask/utils_test.py [hunk 1]
```diff
@@ -75,6 +73,20 @@ def test_get_with_list(self):
...
+             'c': [1, (inc, 1)],
+             'd': [(sum, 'a')],
+             'e': ['a', 'b'],
+             'f': [[[(sum, 'a'), 'c'], (sum, 'b')], 2]}
+        assert self.get(d, 'a') == [1, 2, 3]
+        assert self.get(d, 'b') == [1, 2, 3]
+        assert self.get(d, 'c') == [1, 2]
+        assert self.get(d, 'd') == [6]
+        assert self.get(d, 'e') == [[1, 2, 3], [1, 2, 3]]
+        assert self.get(d, 'f') == [[[6, [1, 2]], 6], 2]
```
Comment (mrocklin) [inline dask/utils_test.py]: Nice test

**Comment by mrocklin** [comment]
I haven't thoroughly reviewed `dask.core.get`.  Everything else looks good to me.

**Comment by jcrist** [comment]
Spec is updated to better reflect the recursive definition. Feedback welcome.

File: docs/source/spec.rst [hunk 2]
```diff
@@ -35,25 +36,27 @@ units of work meant to be run by a single worker. Example:
...
-4.  List of arguments, like ``[1, 'x', (inc, 'x')]``
+3.  A **task** like ``(inc, 'x')`` (see below)
+4.  A list of **computations**, like ``[1, 'x', (inc, 'x')]``
 
-So all of the following are valid tasks:
+So all of the following are valid **computations**:
 
 .. code-block:: python
 
+   np.array([...])
```
Comment (mrocklin) [inline docs/source/spec.rst]: Perhaps start with `123` above `np.array([...])`

Comment (eriknw) [inline docs/source/spec.rst]: What's the purpose of adding `np.array([...])` as an example?  The rest of the examples are computations that have tasks or keys to resolve/compute.  Since a **computation** can contain literals, tasks, or keys, I think it would be best for these examples to expressly exclude literals.

Comment (jcrist) [inline docs/source/spec.rst]: Just giving an example of option 2 above. "computation" was the best general word for the values in a dask graph that I could think of. Happy to change to something simpler like 123, but I do think that a top-level literal should be explicitly shown as valid.

Comment (mrocklin) [inline docs/source/spec.rst]: I agree that we should definitely show that literals are valid.  Perhaps the term "computation" needs to be qualified or replaced here.

Comment (jcrist) [inline docs/source/spec.rst]: The list above does fully enumerate what a computation is, but it may not be the best word. Only other thing I could think of is `value`, but that term is a bit overloaded (could be read as being a literal, or being a value in a dict).

Comment (eriknw) [inline docs/source/spec.rst]: How about **expression**?  An **expression** can be a literal, key, task, or list of expressions.

Comment (mrocklin) [inline docs/source/spec.rst]: I'm probably a bit biased against expression due to history with and perennial comparison with Blaze.  

I think that expression is strongly associated with language things.  This isn't a term in an expression, it's an honest to goodness computation that's about to occur.  We don't know anything semantic about this thing; it's concrete.

Comment (eriknw) [inline docs/source/spec.rst]: I actually like **computation**, but I think we need a clear way to distinguish between **literals** and **computations** that contain keys or tasks that need to be resolved or calculated.

Comment (jcrist) [inline docs/source/spec.rst]: Perhaps switch task and computation around? Tasks can be literals, keys, or lists of tasks. Computations are `(func, *args)`? Otherwise, I can't think of another good option and am ready to punt and just call this fine as is.

Comment (mrocklin) [inline docs/source/spec.rst]: I'm somewhat in favor of punting and going with computation for now.  We can change this in the future if we come up with cleaner ideas.  I'd like to get this merged.

Comment (jcrist) [inline docs/source/spec.rst]: I agree. Good to go as is then?

Comment (mrocklin) [inline docs/source/spec.rst]: +1 from me

**Comment by jcrist** [comment]
Merging.
